rfcs/0157-more-granular-control.md

```diff
@@ -0,0 +1,241 @@
+* The output from `nix-store --export /nix/store/{drv-or-output-path} > output` is not importable using `nix-store --import < output` unless the dependencies were already present on the host. You can't ask --import to look for dependencies in the binary cache. One could get a list of dependency paths using `nix-store --query --requisites /nix/store/{drv-or-output-path}`, but there's no exposed way to effectively "pull" those paths from the binary cache before running --import. The only way is to copy the whole closure for every derivation or its output (from the remote builder to the host).
+* Upstream caching (not caching the packages available in upstream cache such as `https://cache.nixos.org`) requires extra specific 3rd party infrastructure to set up, such as [cachix](https://www.cachix.org) (service) or [attic](https://docs.attic.rs/tutorial.html#pushing) (self-host).
+
+In void (just as a reference), a package that ignores already upstream-dependency is built (in a chroot by default). This output folder can then act as a simple ad-hoc repository mirror with any http server (including `python3 -m http.server -d hostdir/binpkgs` on the host-machine or even github pages after a build from GitHub actions).
+
+
+# Detailed design
+[design]: #detailed-design
+
+While the current configuration does work for most people, and many are content with the existing solutions, a more granular control system would offer tremendous possibilities for customization and make the infrastructure flexible for future shifts.
+
+Instead of deprecating pulseaudio's `sound.*` in favor of `services.pipewire.*`, we should aim to build a flexible API that persists even when new standards emerge, reducing confusion caused by generic names like "sound" being deprecated.
+
+### Everything is opt-in!
+
+First and foremost, during image building (whether it's nixos-rebuild or vm/iso generation), the filesystem should be empty, with no packages or files.
+
+We can then assemble basic components by defining choices such as the bootloader, kernel, initramfs generator [1], libc, init, and then the choice of sound & de/wm.
+
+This approach also facilitates the use of NixOS-modules on non-NixOS nix installations.
+
+<sub>[1]: Recently [dracut merged](https://github.com/NixOS/nixpkgs/issues/77858), but there are no clear instructions on how to use that.</sub>
+
+### Emphasis on flake over shell script wrappers
+
+With the introduction of `nix run` and `nix build`, all current implementations should transition to using these first-party nix commands instead of wrappers like nixos-rebuild, nixos-generate, and home-manager.
+
+Operating everything under flake also opens up various opportunities, such as building the latest nixos iso image with or without plasma/gnome directly from the nixpkgs repository if its derivation is exposed in flake.nix (for example, `nix build nixpkgs#isoInstallerImageWithGnome`).
```

Comment:
This is confusing. _If_ a version of ISO is exposed (like some are in `release.nix`), you can just `nix-build -A` it already. If it is not… then Flakes make things harder because you probably want `nix-build -E ` overriding something from the command line.

Reply:
> Flakes were introduced for more flexibility in pinning, and managing the derivations with more isolation (from NIX_PATH and rest of environment). 

This is not relevant to the alleged features coming from flakes.

> They also allow to be called remotely using github:xxx/yyy and more of such syntax.

This is just telling Nix to download the expressions and evaluate them… which was actually already in use before flakes (at least with full HTTPS urls)

> You can in theory do the same with flakes, and in addition, you can take flake input and write a definition in your flake output which overrides something from original derivation easily, whereas without flakes, how would you import nixos/release.nix in your config and override something without command line option -E through a nix file (to keep the config in vcs).

This is contradictory. Either you can do the same thing with flakes (doubt it, flakes are currently extremely bad at flexibility), or flakes force you to actually create a file and check it in for trying out a small override.

Reply:
What I'm trying to say is that, if it was exposed in flake, we can do something like:

```
outputs = args: {
  my-custom-override = inputs.nixpkgs.nixos.release.imagePath.overrideAttrs (...);
}
```

whereas currently there's no way to access it in nix expressions, as release.nix is not exposed in any way, not even through [default.nix in nixpkgs/nixos](https://github.com/NixOS/nixpkgs/tree/master/nixos).

I'm not sure how would you be editing it via `-E` either, since its not in default.nix, you can't import it as `let a = import <nixpkgs/nixos/release> in a.{path}`...

Reply:
> you can't import it as `let release = import <nixpkgs/nixos/release.nix> {}; in `

Yes you can (with minor fixes to syntax)

> And I don't think flakes are by far any less flexible to do the same thing, rather the enforcement makes them more standardized to work with consistent api.

The mere appearance of `nixpkgs-unfree` shows that flakes are more restrictive (without flakes that single flag is passed via command line or maybe single-command env var; the fact that people do find a ready-made flake that does all the necessary dance useful shows that flake interfaces are not good enough).

Another example is cross-compilation.

Reply:
> Yes you can (with minor fixes to syntax)

Would be great if you could put a complete example, as I'm not familiar enough with old syntax very much :)